### PR TITLE
[Memory Engine] Add MemTabletScan

### PR DESCRIPTION
**Pull request by decster**
This CL add read support(simple projection only full scan) for MemTablet.
It also add a unit test for MemTablet insert-update-scan loop.

Resloves #3733

### CODE REVIEW

File: be/src/olap/memory/mem_tablet_scan.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+    uint64_t limit() const { return _limit; }
+
+    const vector<std::string> columns() const { return _columns; }
+
+private:
+    friend class MemTablet;
+
+    uint64_t _version = UINT64_MAX;
+    uint64_t _limit = UINT64_MAX;
```
Comment (chaoyli) [inline be/src/olap/memory/mem_tablet_scan.h:43]: What's the use of _limit?

File: be/src/olap/memory/mem_tablet.h [hunk 2]
```diff
@@ -79,6 +81,8 @@ class MemTablet : public BaseTablet {
 
     std::mutex _write_lock;
 
+    std::atomic<uint64_t> _latest_version;
```
Comment (chaoyli) [inline be/src/olap/memory/mem_tablet.h:84]: use max_version may be better.

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/olap/base_tablet.cpp [hunk 3]
```diff
@@ -44,11 +43,13 @@ OLAPStatus BaseTablet::set_tablet_state(TabletState state) {
 }
 
 void BaseTablet::_gen_tablet_path() {
-    std::string path = _data_dir->path() + DATA_PREFIX;
-    path = path_util::join_path_segments(path, std::to_string(_tablet_meta->shard_id()));
-    path = path_util::join_path_segments(path, std::to_string(_tablet_meta->tablet_id()));
-    path = path_util::join_path_segments(path, std::to_string(_tablet_meta->schema_hash()));
-    _tablet_path = path;
+    if (_data_dir != nullptr) {
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/olap/base_tablet.cpp:46]: In unittest

File: be/src/olap/memory/mem_tablet_scan.h (hunk 1 above)
Comment (decster) [reply be/src/olap/memory/mem_tablet_scan.h:43]: For example: select * from tablet limit 10;

File: be/src/olap/memory/mem_tablet.h (hunk 2 above)
Comment (decster) [reply be/src/olap/memory/mem_tablet.h]: OK

**Review by chaoyli — APPROVED** [review]

**Pull request merged**